### ISSUE TITLE
The change of the candidate set during training

### ISSUE THREAD

**Issue by leiyu2023** [yandex-research/tabular-dl-tabr#9]
Thank you for your interesting work, it inspires me a lot.There are a couple of questions that have been bugging me
The work initially uses the entire training set as the fixed set of candidates for all objects.
1.The function 'apply_model' removes them. 
2.However,when computing the forward output, it adds the current batch to the candidate set and predicts the output.Should the current batch be added to the candidate set after predicting the output?
3.Additionally, when adding to the candidate set and retrieving context samples, it is guaranteed to retrieve samples that match the target. The related index is then removed from the obtained index.
    I can't understand the role and connection of these three operations, can you provide some suggestions

**Comment by Yura52** [comment]
Sure.

1. In `apply_model`, we have to make sure that the training batch itself is not included in the candidates. Otherwise, for each training object in the batch, the exact copy would be available among the candidates, and the model would simply learn to copy labels, which will not be possible for validation and test data. This risk is applicable only to training data, so candidates are modified only for training batches: [link](https://github.com/yandex-research/tabular-dl-tabr/blob/d628ec7e1c0a66011473021034e7dd4a77740112/bin/tabr.py#L405).

2. The previous step ensures that, **for a training batch**, i-th object in the batch cannot use its copy as a candidate. However, this i-th object also cannot use the (i+1)-th object from the batch as a candidate, which would be totally valid. To restore all possible candidates for all training objects, **for training data**, we put the batch back to candidates: [link](https://github.com/yandex-research/tabular-dl-tabr/blob/d628ec7e1c0a66011473021034e7dd4a77740112/bin/tabr.py#L235). It may seem that this action cancels the effect of the previous step. In fact, this is true! The only difference is that, for each object in the batch, we now know the precise location of its copy among the candidates: the i-th object in the batch is the i-th candidate, because of how candidate_k is constructed: [link](https://github.com/yandex-research/tabular-dl-tabr/blob/d628ec7e1c0a66011473021034e7dd4a77740112/bin/tabr.py#L241).

3. Because of the previous steps, **for a training batch**, it is easy to remove the exact copies from the mined contexts: [link](https://github.com/yandex-research/tabular-dl-tabr/blob/d628ec7e1c0a66011473021034e7dd4a77740112/bin/tabr.py#L272) (specifically, we manually set the infinite distance and remove it in the next line).

It may seem that all of the above is unnecessary, since we always know the precise location of "unfair" candidates for each training batch. And, **for the vanilla setup**, this is true. However, this stops being true when the set of candidates is not equal to the training set (and we conducted such experiments, so it was important to us _not_ to rely on "candidates == training set").


Does this help?

**Comment by leiyu2023** [comment]
Thank you for your reply which solved most of my queries, but I still have some questions.
1.Even if the current object is not removed from the candidate set, it is still easy to find in the distance retrieval phase.The distance is always zero.
2.When placing the batch in front of the candidate set, the position of the same candidate samples as the current object is known, so why is it still necessary to remove it later by setting the distance to infinity and sorting.

**Comment by Yura52** [comment]
> it is still easy to find in the distance retrieval phase.The distance is always zero.

This is not completely true:
- For a target object, there can be a _different_ object among the candidates which is still an exact copy of the target object. And, strictly speaking, without additional business rules, it is fine to use such a candidate as a neighbor.
- Due to numerical instabilities, the distance between two equal objects may not be exactly zero. Then a heuristically selected threshold is needed to filter out potential exact copies. This is a new non-trivial hyperparameter.

To sum up, the current solution may seem a bit complicated, but it is somewhat "straightforward" in a good sense.

> why is it still necessary to remove it later by setting the distance to infinity and sorting.

Because, **for a training object**, this exact object must not be used as a neighbor: otherwise, its label will be used for a prediction [here](https://github.com/yandex-research/tabular-dl-tabr/blob/d628ec7e1c0a66011473021034e7dd4a77740112/bin/tabr.py#L304-L305), which would be a label leak.

**Comment by leiyu2023** [comment]
Thank you very much for your detailed explanation, it has resolved all of my confusion
